### DISCUSSION TITLE
Allow frontend to toggle business rules for resources

### DISCUSSION THREAD

**Discussion by melton-jason** [specify/specify7#5926] [Feature Requests]
> @melton-jason should we trigger business rules on next-tick to run rules only after all changes are applied rather than for each change individually? would that avoid bugs like this?

https://github.com/specify/specify7/pull/4551#pullrequestreview-1886373091

or 

> That may work, although we would still need to be considerate of the order in which the quantityResolved and quantityReturned fields are applied.
> 
> I take it you are suggesting a way to 'skip' the businessrules being run for these specific field sets?
> 
> Besides your proposed solution, here are two more which can also be considered:
> 
> * Allow the `noBusinessRules` option to be passed to `resource.set` and `resource.bulkSet` calls (essentially a more specific implementation of the option which can be passed when a resource is initialized/created or when calling `resource.getRelated`)
>   
>   * `resource.bulkSet({"quantityResolved": 1, "quantityReturned": 1}, {"noBusinessRules": true})`
> * Add a method to `ResourceBase` which accepts a function which would perform all modifications within the function as if business rules were enabled. An example usage might be:
> 
> ```ts
> loanReturn.withoutBusinessRules(() => {
>    loanReturn.set("quantityResolved", 1);
>    loanReturn.set("quantityReturned", 1);
>    // loanReturn quantityResolved is still 1
> )
> // normally this would be
> loanReturn.set("quantityResolved", 1);
> loanReturn.set("quantityReturned", 1);
> // loanReturn quantityResolved is 2
> ```
> 
> For both scenarios some modifications would need to be made to both businessRules.ts and ResourceBase. Currently, the only way to disable rules is on resource initialization. See
> 
> https://github.com/specify/specify7/blob/7041601872f434f61b524f392a9bcf8f2f86d5b8/specifyweb/frontend/js_src/lib/components/DataModel/resourceApi.js#L118

https://github.com/specify/specify7/pull/4551#issuecomment-1952876220

In short, the frontend should be allowed to ignore businessrules for one or more changes/operations for a resource and subsequently re-enable business for the resource once the desired operations are performed

**Comment by maxpatiiuk** [comment]
> That may work, although we would still need to be considerate of the order in which the quantityResolved and quantityReturned fields are applied.
> I take it you are suggesting a way to 'skip' the businessrules being run for these specific field sets?

I might have misunderstood the problem, but my impression was that because each business rule change was considered individually, a given business rule change was applied twice, causing incorrect count

my idea was that all field changes would schedule business rules, and then business rules would run on the next tick using `queueMicrotask`. and then, we could potentially say inside the definition for the quantityResolved rule: "if quantityReturned wasn't modified in the same batch of modifications, only then run this business rule"

when calling .save(), it should also await that there are no more pending business rules (I'm not sure if it does that right now ❗)

---

I like this other approach you suggested for it's simplicity:

```ts
loanReturn.withoutBusinessRules(() => {
   loanReturn.set("quantityResolved", 1);
   loanReturn.set("quantityReturned", 1);
   // loanReturn quantityResolved is still 1
)
```

though, my issue with it is that it shifts the responsibility for knowing what business rules exist and in what order they run from the business rule writer to the developer of any component that modifies any database field.
add to it the possibility of adding new business rules, or ability for admins to configure business rules, and the chance for bugs and mistakes goes up

if we could figure out a way to have business rules "just work" with any modifications you make, that would be nice.

---

maybe we should rethink what business rules are and when they are run - i.e should they only be run when user explicitly makes a change though the form? and in other places require opt-in? or when defining a business rule, you should say whether business rule should always be run or only on form changes?
i.e are business rules run right now for workbench uploads? would they run for bulk edits in the future? should they?

---

in general, business rule deal with data integrity, which, needless to say is really important, more so than most of the other things we deal with on the front-end. so coming up with a "pit of success" for business rules, where without knowing what business rules exist (or what will exist, or that they exist at all), things would correctly
